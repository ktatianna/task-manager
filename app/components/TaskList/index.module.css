.list {
  list-style: none;
  margin: 0 auto;
  padding: 0;
  width: 400px;
}

.emptyMessage {
  color: #6b7280;
  font-size: 18px;
  margin: 32px 0;
  text-align: center;
}

.clearCompletedBtn {
  display: block;
  margin: 0 auto 1.5rem auto;
  padding: 0.5rem 1.5rem;
  background: transparent;
  color: #7c3aed;
  border-radius: 6px;
  font-size: 1rem;
  cursor: pointer;
  transition: background 0.2s, color 0.2s;
}
.clearCompletedBtn:hover {
  background: #ededed;
  color: #5b21b6;
  border-color: #5b21b6;
}