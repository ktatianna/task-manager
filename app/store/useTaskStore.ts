import { create } from "zustand";
import { Task } from "../types/task";
import { v4 as uuidv4 } from "uuid";
import { saveTasksToStorage } from "../utils/storage";

interface TaskStore {
  tasks: Task[];
  setTasks: (tasks: Task[]) => void;
  addTask: (description: string) => void;
  toggleTask: (id: string) => void;
  deleteTask: (id: string) => void;
  clearCompleted: () => void;
}

export const useTaskStore = create<TaskStore>((set) => ({
  tasks: [],
  setTasks: (tasks) => set({ tasks }),
  addTask: (description) =>
    set((state) => {
      const newTasks = [...state.tasks, { id: uuidv4(), description, completed: false }];
      saveTasksToStorage(newTasks);
      return { tasks: newTasks };
    }),
  toggleTask: (id) =>
    set((state) => {
      const newTasks = state.tasks.map((task) =>
        task.id === id ? { ...task, completed: !task.completed } : task
      );
      saveTasksToStorage(newTasks);
      return { tasks: newTasks };
    }),
  deleteTask: (id) =>
    set((state) => {
      const newTasks = state.tasks.filter((task) => task.id !== id);
      saveTasksToStorage(newTasks);
      return { tasks: newTasks };
    }),
  clearCompleted: () =>
    set((state) => {
      const newTasks = state.tasks.filter((task) => !task.completed);
      saveTasksToStorage(newTasks);
      return { tasks: newTasks };
    }),
}));
