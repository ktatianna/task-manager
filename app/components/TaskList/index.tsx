'use client'

import { useEffect, useState } from "react"
import { useTaskStore } from "@/app/store/useTaskStore"
import { loadTasksFromStorage } from "@/app/utils/storage"
import TaskItem from "../TaskItem"
import styles from "./index.module.css"
import EmptyState from "../EmptyState"

const TaskList = () => {
  const tasks = useTaskStore((state) => state.tasks)
  const setTasks = useTaskStore((state) => state.setTasks)
  const clearCompleted = useTaskStore((state) => state.clearCompleted)
  const hasCompleted = tasks.some(task => task.completed)

  const [isLoaded, setIsLoaded] = useState(false)

  useEffect(() => {
    const tasksFromStorage = loadTasksFromStorage();
    setTasks(tasksFromStorage)
    setIsLoaded(true)
  }, [setTasks]);

  if (!isLoaded) {
    return <div  className={styles.loading}>Loading tasks...</div>
  }

  if (tasks.length === 0) {
    return <EmptyState />
  }

  return (
    <>
      <ul className={styles.list}>
        {tasks.map((task) => (
          <TaskItem
            key={task.id}
            task={task}
          />
        ))}
      </ul>
      {hasCompleted && (
        <button
          onClick={clearCompleted}
          className={styles.clearCompletedBtn}
        >
          Delete completed tasks
        </button>
      )}
    </>
  )
}

export default TaskList;