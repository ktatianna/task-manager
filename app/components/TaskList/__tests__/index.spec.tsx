import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { render, screen } from "@testing-library/react";
import TaskList from "../index";
import { useTaskStore } from "@/app/store/useTaskStore";

jest.mock("@/app/store/useTaskStore", () => ({
  useTaskStore: jest.fn(),
}));

const defaultTasks = [
  { id: "1", description: "Leer un libro", completed: false },
  { id: "2", description: "Hacer ejercicio", completed: true },
];

const setup = (tasksOverride = defaultTasks, clearCompletedMock = jest.fn()) => {
  (useTaskStore as unknown as jest.Mock).mockImplementation((selector) =>
    selector({ tasks: tasksOverride, clearCompleted: clearCompletedMock })
  );
  return render(<TaskList />);
};

describe("TaskList", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should match the snapshot", () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("should render all tasks", () => {
    setup();

    expect(screen.getByText("Leer un libro")).toBeInTheDocument();
    expect(screen.getByText("Hacer ejercicio")).toBeInTheDocument();
  });

  it("should render empty message if there are no tasks", () => {
    setup([]);
    expect(screen.getByText("You have no tasks yet")).toBeInTheDocument();
  });

  it("should call clearCompleted when the button is clicked", async () => {
    const clearCompletedMock = jest.fn();
    setup(defaultTasks, clearCompletedMock);

    const button = screen.getByRole("button", { name: "Delete completed tasks" });
    await userEvent.click(button);
    expect(clearCompletedMock).toHaveBeenCalled();
  });
});
