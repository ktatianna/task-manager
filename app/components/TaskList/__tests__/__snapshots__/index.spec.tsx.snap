// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList should match the snapshot 1`] = `
<div>
  <ul
    class="list"
  >
    <li
      class="task "
    >
      <div
        class="taskContent"
      >
        <input
          aria-label="Leer un libro"
          class="checkbox"
          id="1"
          type="checkbox"
        />
        <span
          class="description"
        >
          Leer un libro
        </span>
      </div>
      <button
        aria-label="Delete task"
        class="deleteButton"
      >
        <svg
          fill="currentColor"
          height="20"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m112 112 20 320c.95 18.49 14.4 32 32 32h184c17.67 0 30.87-13.51 32-32l20-320"
            fill="none"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="32"
          />
          <path
            d="M80 112h352"
            stroke-linecap="round"
            stroke-miterlimit="10"
            stroke-width="32"
          />
          <path
            d="M192 112V72h0a23.93 23.93 0 0 1 24-24h80a23.93 23.93 0 0 1 24 24h0v40m-64 64v224m-72-224 8 224m136-224-8 224"
            fill="none"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="32"
          />
        </svg>
      </button>
    </li>
    <li
      class="task completed"
    >
      <div
        class="taskContent"
      >
        <input
          aria-label="Hacer ejercicio"
          checked=""
          class="checkbox"
          id="2"
          type="checkbox"
        />
        <span
          class="description"
        >
          Hacer ejercicio
        </span>
      </div>
      <button
        aria-label="Delete task"
        class="deleteButton"
      >
        <svg
          fill="currentColor"
          height="20"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m112 112 20 320c.95 18.49 14.4 32 32 32h184c17.67 0 30.87-13.51 32-32l20-320"
            fill="none"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="32"
          />
          <path
            d="M80 112h352"
            stroke-linecap="round"
            stroke-miterlimit="10"
            stroke-width="32"
          />
          <path
            d="M192 112V72h0a23.93 23.93 0 0 1 24-24h80a23.93 23.93 0 0 1 24 24h0v40m-64 64v224m-72-224 8 224m136-224-8 224"
            fill="none"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="32"
          />
        </svg>
      </button>
    </li>
  </ul>
  <button
    class="clearCompletedBtn"
  >
    Delete completed tasks
  </button>
</div>
`;
